src/mono/wasm/templates/templates/browser/app-support.js

```diff
@@ -0,0 +1,427 @@
+        },
+        preRun: () => {
+            if (!processedArguments.enable_gc) {
+                INTERNAL.mono_wasm_enable_on_demand_gc(0);
+            }
+        },
+        onDotnetReady: () => {
+            let wds = Module.FS.stat(processedArguments.working_dir);
+            if (wds === undefined || !Module.FS.isDir(wds.mode)) {
+                set_exit_code(1, `Could not find working directory ${processedArguments.working_dir}`);
+                return;
+            }
+
+            Module.FS.chdir(processedArguments.working_dir);
+
+            if (processedArguments.runtime_args.length > 0)
+                INTERNAL.mono_wasm_set_runtime_options(processedArguments.runtime_args);
+
+            console.info("Initializing.....");
+            Object.assign(App, { MONO, INTERNAL, BINDING, Module, processedArguments });
+
+            try {
+                if (App.init)
+                {
+                    let ret = App.init();
+                    Promise.resolve(ret).then(function (code) { set_exit_code(code ?? 0); });
+                }
+                else
+                {
+                    console.log("WASM ERROR: no App.init defined");
+                    set_exit_code(1, "WASM ERROR: no App.init defined");
+                }
+            } catch (err) {
+                console.log(`WASM ERROR ${err}`);
+                if (is_browser && document.getElementById("out"))
+                    document.getElementById("out").innerHTML = `error: ${err}`;
+                set_exit_code(1, err);
+            }
+        },
+        onAbort: (error) => {
+            set_exit_code(1, stringify_as_error_with_stack(new Error()));
+        },
+    }))
+}).catch(function (err) {
+    set_exit_code(1, "failed to load the dotnet.js file.\n" + stringify_as_error_with_stack(err));
+});
+
+function set_exit_code(exit_code, reason) {
+    if (reason) {
+        if (reason instanceof Error)
+            console.error(stringify_as_error_with_stack(reason));
+        else if (typeof reason == "string")
+            console.error(reason);
+        else
+            console.error(JSON.stringify(reason));
+    }
+
+    if (is_browser) {
+        if (App.Module) {
+            // Notify the selenium script
+            App.Module.exit_code = exit_code;
+        }
+
+        //Tell xharness WasmBrowserTestRunner what was the exit code
+        const tests_done_elem = document.createElement("label");
+        tests_done_elem.id = "tests_done";
+        tests_done_elem.innerHTML = exit_code.toString();
+        document.body.appendChild(tests_done_elem);
+
+        if (forward_console)  {
+            const stop_when_ws_buffer_empty = () => {
+                if (consoleWebSocket.bufferedAmount == 0) {
+                    // tell xharness WasmTestMessagesProcessor we are done. 
+                    // note this sends last few bytes into the same WS
+                    console.log("WASM EXIT " + exit_code);
```

Comment:
Is exit code "in-scope" for browser host ?

Reply:
Some of this stuff can be pruned like you said. But a big portion of the code is shared, like the arguments processing etc. It will be used in 3 places - the two templates, and test-main.js .

Reply:
> Is exit code "in-scope" for browser host ?

This is actually useful in the case when you use `--forward-console`, which shows the messages in the terminal. In that case, the exit message is helpful, and I'm able to use it for tests too. So, I'm printing that only when forwarding the console.